File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int res = 0;
        Stack<Integer> stack = new Stack<>();
        stack.push(-1);
        for(int i = 0; i < heights.length; i++) {
            while(stack.peek() != -1 && heights[stack.peek()] > heights[i]) {
                int h = heights[stack.pop()];
                int w = i - stack.peek() - 1;
                res = Math.max(res, w * h);
            }
            stack.push(i);
        }
        while(stack.peek() != -1) {
            int h = heights[stack.pop()];
            int w = heights.length - stack.peek() - 1;
            res = Math.max(res, w * h);
        }
        return res;
    }
}

File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        res = 0
        stack = []
        stack.append(-1)
        n = len(heights)
        for i in range(len(heights)):
            while stack[-1] != -1 and heights[i] < heights[stack[-1]]:
                l = stack.pop()
                res = max(res, (i - stack[-1] - 1) * heights[l])
                
            stack.append(i)
        while stack[-1] != -1:
            l = stack.pop()
            res = max(res, (n - stack[-1] - 1) * (heights[l]))
            
            
        return res
            